File: google/main/src/core/query/Terms.java
package main.src.core.query;

import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.*;

import main.src.core.engine.InvertedIndex;
import main.src.core.structures.Token;
import main.src.core.structures.Document;

public abstract class Terms {

    protected ArrayList<Token> keys;

    public Terms() {
        this.keys = new ArrayList<>();
    }

    public ArrayList<Token> getKeys() {
        return keys;
    }

    protected void collect(String expression, Pattern pattern, int regexGroupIndex) {
        Matcher matcher = pattern.matcher(expression);
        while (matcher.find())
            this.keys.add(new Token(matcher.group(regexGroupIndex)));
    }

    public boolean isEmpty() {
        return this.keys.isEmpty();
    }

    public abstract HashSet<Document> getResults(InvertedIndex index);
}

File: GoogleSharp/GoogleSharp/src/core/query/Terms.cs
// Standard Library
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

// Internal
using GoogleSharp.Src.Core.Engine;
using GoogleSharp.Src.Core.Structures;

namespace GoogleSharp.Src.Core.Query
{
    public abstract class Terms
    {
        public List<Token> Tokens { get; set; }

        public Terms()
        {
            Tokens = new List<Token>();
        }

        public abstract HashSet<Document> GetResults(IInvertedIndex index);

        public bool IsEmpty()
        {
            return !Tokens.Any();
        }

        protected void Collect(string expression, Regex pattern, int regexGroupIndex)
        {
            foreach (var match in pattern.Matches(expression).OfType<Match>())
                Tokens.Add(new Token(match.Groups[regexGroupIndex].Value));
        }
    }
}
